### ISSUE TITLE
6.4.0.161 breaks Monodevelop

### ISSUE THREAD

**Issue by Commaster** [mono/mono#16166]
## Steps to Reproduce

1. Update mono from 6.4.0.159 to 6.4.0.161
2. Launch the latest Monodevelop (7.8.4.2)
3. Observe

### Current Behavior

All File-related operations seem to be broken.

### Expected Behavior

Normal project loading

## On which platforms did you notice this

[ ] macOS
[x] Linux
[ ] Windows

**Version Used**:

Mono 6.4.0.159 (tarball Thu Aug  8 10:34:21 UTC 2019) (64-bit)
Mono 6.4.0.161 (tarball Fri Aug  9 10:18:10 UTC 2019) (64-bit)
MonoDevelop 7.8.4 (build 2)

## Stacktrace

```
ERROR [2019-08-09 19:49:09Z]: Error while parsing .pc file: /usr/lib/pkgconfig/mono-nunit.pc
System.IO.FileNotFoundException: Could not find file "/usr/lib/pkgconfig/mono-nunit.pc"
File name: '/usr/lib/pkgconfig/mono-nunit.pc'
  at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.Boolean anonymous, System.IO.FileOptions options) <0x7f07492a3c40 + 0x003e9> in <8ba1217e6efa47ac8058121a30b66184>:0 
  at System.IO.FileStream..ctor (System.String path, System.IO.FileMode mode, System.IO.FileAccess access, System.IO.FileShare share, System.Int32 bufferSize, System.IO.FileOptions options) <0x7f07492a37a0 + 0x00072> in <8ba1217e6efa47ac8058121a30b66184>:0 
  at (wrapper remoting-invoke-with-check) System.IO.FileStream..ctor(string,System.IO.FileMode,System.IO.FileAccess,System.IO.FileShare,int,System.IO.FileOptions)
  at System.IO.StreamReader..ctor (System.String path, System.Text.Encoding encoding, System.Boolean detectEncodingFromByteOrderMarks, System.Int32 bufferSize) <0x7f074926ad80 + 0x000cd> in <8ba1217e6efa47ac8058121a30b66184>:0 
  at System.IO.StreamReader..ctor (System.String path, System.Boolean detectEncodingFromByteOrderMarks) <0x7f074926ac60 + 0x00048> in <8ba1217e6efa47ac8058121a30b66184>:0 
  at System.IO.StreamReader..ctor (System.String path) <0x7f074926ac10 + 0x00037> in <8ba1217e6efa47ac8058121a30b66184>:0 
  at (wrapper remoting-invoke-with-check) System.IO.StreamReader..ctor(string)
  at Mono.PkgConfig.PcFile.Load (System.String pcfile) [0x00020] in <f7bdc01ed9fa47b588f77ebb8789f22c>:0 
  at Mono.PkgConfig.PcFileCache`1[TP].ParsePackageInfo (System.String pcfile) [0x00007] in <f7bdc01ed9fa47b588f77ebb8789f22c>:0 
  at Mono.PkgConfig.PcFileCache`1[TP].GetPackageInfo (System.String file) [0x00074] in <f7bdc01ed9fa47b588f77ebb8789f22c>:0 
INFO [2019-08-09 19:49:09Z]: [464 (+85)] END: Target runtimes loaded
```

```
FATAL ERROR [2019-08-09 19:49:14Z]: Can't create roslyn workspace
System.InvalidOperationException: Sequence contains no elements
  at System.Linq.Enumerable.First[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x00010] in <6ab3b7e94c2a40f68ec7761a701ea26e>:0 
  at Microsoft.VisualStudio.Composition.PartDiscovery.GetElementTypeFromMany (System.Type type) [0x000e9] in <0f47189cc3a34b52b8c3c4e5dc0e7ed3>:0 
  at Microsoft.VisualStudio.Composition.RuntimeComposition+RuntimeImport.get_ImportingSiteTypeWithoutCollection () [0x00026] in <0f47189cc3a34b52b8c3c4e5dc0e7ed3>:0 
  at Microsoft.VisualStudio.Composition.RuntimeComposition+RuntimeImport.get_IsLazy () [0x0000d] in <0f47189cc3a34b52b8c3c4e5dc0e7ed3>:0 
  at Microsoft.VisualStudio.Composition.RuntimeComposition+RuntimeImport.get_LazyFactory () [0x00008] in <0f47189cc3a34b52b8c3c4e5dc0e7ed3>:0 
  at Microsoft.VisualStudio.Composition.RuntimeExportProviderFactory+RuntimeExportProvider.GetValueForImportSite (Microsoft.VisualStudio.Composition.RuntimeExportProviderFactory+RuntimeExportProvider+RuntimePartLifecycleTracker importingPartTracker, Microsoft.VisualStudio.Composition.RuntimeComposition+RuntimeImport import) [0x0000c] in <0f47189cc3a34b52b8c3c4e5dc0e7ed3>:0 
  at Microsoft.VisualStudio.Composition.RuntimeExportProviderFactory+RuntimeExportProvider+RuntimePartLifecycleTracker.<CreateValue>b__10_0 (Microsoft.VisualStudio.Composition.RuntimeComposition+RuntimeImport import) [0x00006] in <0f47189cc3a34b52b8c3c4e5dc0e7ed3>:0 
  at System.Linq.Enumerable+SelectArrayIterator`2[TSource,TResult].ToArray () [0x00012] in <6ab3b7e94c2a40f68ec7761a701ea26e>:0 
  at System.Linq.Enumerable.ToArray[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x0001f] in <6ab3b7e94c2a40f68ec7761a701ea26e>:0 
  at Microsoft.VisualStudio.Composition.RuntimeExportProviderFactory+RuntimeExportProvider+RuntimePartLifecycleTracker.CreateValue () [0x0006a] in <0f47189cc3a34b52b8c3c4e5dc0e7ed3>:0 
  at Microsoft.VisualStudio.Composition.ExportProvider+PartLifecycleTracker.Create () [0x000c3] in <0f47189cc3a34b52b8c3c4e5dc0e7ed3>:0 
  at Microsoft.VisualStudio.Composition.ExportProvider+PartLifecycleTracker.MoveNext (Microsoft.VisualStudio.Composition.ExportProvider+PartLifecycleState nextState) [0x00042] in <0f47189cc3a34b52b8c3c4e5dc0e7ed3>:0 
  at Microsoft.VisualStudio.Composition.ExportProvider+PartLifecycleTracker.MoveToState (Microsoft.VisualStudio.Composition.ExportProvider+PartLifecycleState requiredState) [0x00008] in <0f47189cc3a34b52b8c3c4e5dc0e7ed3>:0 
  at Microsoft.VisualStudio.Composition.ExportProvider+PartLifecycleTracker.GetValueReadyToExpose () [0x00026] in <0f47189cc3a34b52b8c3c4e5dc0e7ed3>:0 
  at Microsoft.VisualStudio.Composition.RuntimeExportProviderFactory+RuntimeExportProvider+<>c__DisplayClass15_0.<GetExportedValueHelper>b__0 () [0x00108] in <0f47189cc3a34b52b8c3c4e5dc0e7ed3>:0 
  at Microsoft.VisualStudio.Composition.RuntimeExportProviderFactory+RuntimeExportProvider.GetValueForImportElement (Microsoft.VisualStudio.Composition.RuntimeExportProviderFactory+RuntimeExportProvider+RuntimePartLifecycleTracker importingPartTracker, Microsoft.VisualStudio.Composition.RuntimeComposition+RuntimeImport import, Microsoft.VisualStudio.Composition.RuntimeComposition+RuntimeExport export, System.Func`3[T1,T2,TResult] lazyFactory) [0x000ae] in <0f47189cc3a34b52b8c3c4e5dc0e7ed3>:0 
  at Microsoft.VisualStudio.Composition.RuntimeExportProviderFactory+RuntimeExportProvider.GetValueForImportSite (Microsoft.VisualStudio.Composition.RuntimeExportProviderFactory+RuntimeExportProvider+RuntimePartLifecycleTracker importingPartTracker, Microsoft.VisualStudio.Composition.RuntimeComposition+RuntimeImport import) [0x00250] in <0f47189cc3a34b52b8c3c4e5dc0e7ed3>:0 
  at Microsoft.VisualStudio.Composition.RuntimeExportProviderFactory+RuntimeExportProvider+RuntimePartLifecycleTracker.<CreateValue>b__10_0 (Microsoft.VisualStudio.Composition.RuntimeComposition+RuntimeImport import) [0x00006] in <0f47189cc3a34b52b8c3c4e5dc0e7ed3>:0 
  at System.Linq.Enumerable+SelectArrayIterator`2[TSource,TResult].ToArray () [0x00012] in <6ab3b7e94c2a40f68ec7761a701ea26e>:0 
  at System.Linq.Enumerable.ToArray[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x0001f] in <6ab3b7e94c2a40f68ec7761a701ea26e>:0 
  at Microsoft.VisualStudio.Composition.RuntimeExportProviderFactory+RuntimeExportProvider+RuntimePartLifecycleTracker.CreateValue () [0x0006a] in <0f47189cc3a34b52b8c3c4e5dc0e7ed3>:0 
  at Microsoft.VisualStudio.Composition.ExportProvider+PartLifecycleTracker.Create () [0x000c3] in <0f47189cc3a34b52b8c3c4e5dc0e7ed3>:0 
  at Microsoft.VisualStudio.Composition.ExportProvider+PartLifecycleTracker.MoveNext (Microsoft.VisualStudio.Composition.ExportProvider+PartLifecycleState nextState) [0x00042] in <0f47189cc3a34b52b8c3c4e5dc0e7ed3>:0 
  at Microsoft.VisualStudio.Composition.ExportProvider+PartLifecycleTracker.MoveToState (Microsoft.VisualStudio.Composition.ExportProvider+PartLifecycleState requiredState) [0x00008] in <0f47189cc3a34b52b8c3c4e5dc0e7ed3>:0 
  at Microsoft.VisualStudio.Composition.ExportProvider+PartLifecycleTracker.GetValueReadyToExpose () [0x00026] in <0f47189cc3a34b52b8c3c4e5dc0e7ed3>:0 
  at Microsoft.VisualStudio.Composition.ExportProvider+<>c__DisplayClass53_0.<CreateExport>b__0 () [0x0002e] in <0f47189cc3a34b52b8c3c4e5dc0e7ed3>:0 
  at System.Lazy`1[T].ViaFactory (System.Threading.LazyThreadSafetyMode mode) <0x7f0748fc8ad0 + 0x00166> in <8ba1217e6efa47ac8058121a30b66184>:0 
  at System.Lazy`1[T].ExecutionAndPublication (System.LazyHelper executionAndPublication, System.Boolean useDefaultConstructor) <0x7f0748fc8c40 + 0x00097> in <8ba1217e6efa47ac8058121a30b66184>:0 
  at System.Lazy`1[T].CreateValue () <0x7f0748fc8f50 + 0x000a4> in <8ba1217e6efa47ac8058121a30b66184>:0 
  at System.Lazy`1[T].get_Value () <0x7f0748fc9170 + 0x0003e> in <8ba1217e6efa47ac8058121a30b66184>:0 
  at Microsoft.VisualStudio.Composition.Export.get_Value () [0x00000] in <0f47189cc3a34b52b8c3c4e5dc0e7ed3>:0 
  at Microsoft.VisualStudio.Composition.NetFxAdapters+MefV1ExportProvider+<>c__DisplayClass12_0.<UnwrapExport>b__1 () [0x00000] in <0f47189cc3a34b52b8c3c4e5dc0e7ed3>:0 
  at System.ComponentModel.Composition.Primitives.Export.GetExportedValueCore () [0x00008] in <188bfa3e72384046a22fe87e2c9da22a>:0 
  at System.ComponentModel.Composition.Primitives.Export.get_Value () [0x0000f] in <188bfa3e72384046a22fe87e2c9da22a>:0 
  at System.ComponentModel.Composition.ExportServices.GetCastedExportedValue[T] (System.ComponentModel.Composition.Primitives.Export export) [0x00006] in <188bfa3e72384046a22fe87e2c9da22a>:0 
  at System.ComponentModel.Composition.ExportServices+<>c__DisplayClass11_0`2[T,M].<CreateStronglyTypedLazyOfTM>b__1 () [0x00000] in <188bfa3e72384046a22fe87e2c9da22a>:0 
  at System.Lazy`1[T].PublicationOnlyViaFactory (System.LazyHelper initializer) <0x7f0748fc8e60 + 0x0003e> in <8ba1217e6efa47ac8058121a30b66184>:0 
  at System.Lazy`1[T].CreateValue () <0x7f0748fc8f50 + 0x00063> in <8ba1217e6efa47ac8058121a30b66184>:0 
  at System.Lazy`1[T].get_Value () <0x7f0748fc9170 + 0x0003e> in <8ba1217e6efa47ac8058121a30b66184>:0 
  at Microsoft.CodeAnalysis.Host.Mef.MefWorkspaceServices+<>c__DisplayClass5_0.<.ctor>b__1 () [0x00000] in <5da55b0063ab4cc089c71dd7627c0eb2>:0 
  at System.Lazy`1[T].ViaFactory (System.Threading.LazyThreadSafetyMode mode) <0x7f0748fc8ad0 + 0x00166> in <8ba1217e6efa47ac8058121a30b66184>:0 
  at System.Lazy`1[T].ExecutionAndPublication (System.LazyHelper executionAndPublication, System.Boolean useDefaultConstructor) <0x7f0748fc8c40 + 0x00097> in <8ba1217e6efa47ac8058121a30b66184>:0 
  at System.Lazy`1[T].CreateValue () <0x7f0748fc8f50 + 0x000a4> in <8ba1217e6efa47ac8058121a30b66184>:0 
  at System.Lazy`1[T].get_Value () <0x7f0748fc9170 + 0x0003e> in <8ba1217e6efa47ac8058121a30b66184>:0 
  at Microsoft.CodeAnalysis.Host.Mef.MefWorkspaceServices.GetService[TWorkspaceService] () [0x00014] in <5da55b0063ab4cc089c71dd7627c0eb2>:0 
  at Microsoft.CodeAnalysis.Workspace..ctor (Microsoft.CodeAnalysis.Host.HostServices host, System.String workspaceKind) [0x0009e] in <5da55b0063ab4cc089c71dd7627c0eb2>:0 
  at MonoDevelop.Ide.TypeSystem.MonoDevelopWorkspace..ctor (MonoDevelop.Projects.Solution solution) [0x00077] in <0d5fa71f3b5e459095de129b1f163448>:0 
  at MonoDevelop.Ide.TypeSystem.TypeSystemService..cctor () [0x000c0] in <0d5fa71f3b5e459095de129b1f163448>:0 
```

Please restore 6.4.0.159 in the preview-\<version\> repository or fix this issue.

**Comment by marek-safar** [comment]
@steveisok this looks like #16054 fix needs more work

**Comment by steveisok** [comment]
So 6.4.0.161 had a bad commit https://github.com/mono/mono/commit/893099f23fdf3881d502f000c3a7942f7f04a802

And was fixed by https://github.com/mono/mono/commit/a7919897e8bc1268547d60ca2283e619a3b3cc45

I tried MD w/ 6.4.165 (Latest preview) and MD worked as expected. 

@Commaster can you please update to 6.4.165 or later and try again?

**Comment by Commaster** [comment]
@steveisok Thank you very much, I can open files with 6.4.0.145 again.
Had to suffer on 6.0.0.319 for just the weekend.
👍

**Comment by Commaster** [comment]
Okay, this was weird.

Tried it on my work PC:
Updated from 159 to 165 directly -> broken MD.
Purged and reinstalled 165 -> broken MD.
Purged 165, installed 6.0.0.319, ran MD with that for a bit, updated to 6.4.0.165 -> now it's working.

What happened? Why is the jump through mono 6.0.x mandatory?

**Comment by akoeplinger** [comment]
Yeah that's definitely weird. Can you try to make a diff of the files when you install 6.4.0.165 fresh vs. when you upgrade it from 6.0.0.319?

**Comment by Commaster** [comment]
Sad news. I can't re-break MD by purging and reinstalling 165. Guess that jump through 6.0.0.319 fixed something permanently... I suspect some issues from earlier 6.4.0.x versions piled up and only 6.0.0.x would clean them up.

**Comment by johnbester** [comment]
This problem still exists on Ubuntu 20.04, but I found a workaround: (mono 6.12 seems to break something)
1. Get the latest version of mono without any additional repositories (in my case it is 6.8.0.105+dfsg-2)
2. Install mono-complete before adding new repositories
apt update
apt install mono
3. Now add the repository to install monodevelop (at this time stable-focal does not contain monodevelop)
deb [arch=amd64] https://download.mono-project.com/repo/ubuntu stable-bionic main
4. Install monodevelop
apt update
apt install monodevelop
5. Disable the mono repository before doing any upgrades and do apt update again

I am attaching a script which I used to force a version. It does perform the installation correctly, but the versions are not fixed for some reason and apt upgrade will replace mono packages, so disabling the repository is important.
[mono-downgrade.zip](https://github.com/mono/mono/files/5348677/mono-downgrade.zip)
